src/frontend/flight/sync.rs

```diff
@@ -0,0 +1,620 @@
+    }
+}
+
+fn now() -> u64 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_secs()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::context::test_utils::in_memory_context;
+    use crate::frontend::flight::sync::{SeafowlDataSyncManager, SequenceNumber};
+    use crate::frontend::flight::SEAFOWL_SYNC_DATA_UD_FLAG;
+    use arrow::{array::RecordBatch, util::data_gen::create_random_batch};
+    use arrow_schema::{DataType, Field, Schema};
+    use rand::Rng;
+    use rstest::rstest;
+    use std::collections::VecDeque;
+    use std::sync::Arc;
+
+    // Create a randomly sized vector of random record batches with
+    // a pre-defined schema
+    fn random_batches() -> Vec<RecordBatch> {
+        let schema = Arc::new(Schema::new(vec![
+            Field::new("c1", DataType::Int32, true),
+            Field::new("c2", DataType::Utf8, true),
+            Field::new(SEAFOWL_SYNC_DATA_UD_FLAG, DataType::Boolean, false),
+        ]));
+
+        // Generate a random length between 1 and 3
+        let len: usize = rand::thread_rng().gen_range(1..=3);
+
+        (0..len)
+            .map(|_| create_random_batch(schema.clone(), 10, 0.2, 0.8).unwrap())
+            .collect()
+    }
+
+    fn assert_sequences(
+        sync_mgr: &mut SeafowlDataSyncManager,
+        origin: &String,
+        memory_sequence: Option<u64>,
+        durable_sequence: Option<u64>,
+    ) {
+        assert_eq!(
+            sync_mgr.stored_sequences(origin),
+            (memory_sequence, durable_sequence),
+            "Unexpected memory/durable sequence; \nseqs {:?}",
+            sync_mgr.seqs,
+        )
+    }
+
+    const FLUSH: (&str, i64) = ("__flush", -1);
+
+    #[rstest]
+    #[case::basic(
+        &[("table_1", 1), ("table_2", 2), ("table_1", 3), FLUSH, FLUSH],
+        vec![Some(1), Some(3)]
+    )]
+    #[case::doc_example(
+        &[("table_1", 1), ("table_2", 1), ("table_3", 2), ("table_1", 3), ("table_2", 3), FLUSH, FLUSH, FLUSH],
+        vec![None, Some(1), Some(3)]
+    )]
+    #[case::long_sequence(
+        &[("table_1", 1), ("table_1", 1), ("table_1", 1), ("table_1", 1), ("table_2", 1),
+            ("table_2", 2), ("table_2", 2), ("table_2", 2), ("table_3", 2), ("table_3", 3),
+            ("table_3", 3), ("table_1", 4), ("table_3", 4), ("table_1", 4), ("table_3", 4),
+            FLUSH, FLUSH, FLUSH],
+        vec![None, Some(1), Some(4)]
+    )]
+    #[case::long_sequence_mid_flush(
+        &[("table_1", 1), ("table_1", 1), ("table_1", 1), FLUSH, ("table_1", 1), ("table_2", 1),
+            ("table_2", 2), ("table_2", 2), FLUSH, ("table_2", 2), ("table_3", 2), FLUSH, ("table_3", 3),
+            FLUSH, ("table_3", 3), ("table_1", 4), ("table_3", 4), ("table_1", 4), FLUSH, ("table_3", 4),
+            FLUSH, FLUSH],
+        // Reasoning for the observed durable sequences:
+        // - seq 1 not seen last sync
+        // - seq 1 seen last sync, but it is in a unflushed table (2)
```

Comment:
I got confused as to why this is None and not 1, but it's because after the previous flush, we received data for table_1 and table_2, and the first flush after that only flushes table_1?

Reply:
That is correct, each `FLUSH` in the test executes a single flush call, which translates to a single table location with the oldest pending changes.